src/VisualStudio/Core/Def/Implementation/ProjectSystem/DocumentProvider.cs

```diff
@@ -207,51 +331,78 @@ public void NotifyDocumentRegisteredToProject(IVisualStudioHostDocument document
             uint docCookie;
             if (RunningDocumentTable.TryGetCookieForInitializedDocument(document.Key.Moniker, out docCookie))
             {
-                TryProcessOpenForDocCookie(docCookie);
+                TryProcessOpenForDocCookie(docCookie, cancellationToken);
             }
+
+            cancellationToken.ThrowIfCancellationRequested();
+            RemovePendingDocumentInitializationTask(document);
         }
 
-        private void TryProcessOpenForDocCookie(uint docCookie)
+        private void TryProcessOpenForDocCookie(uint docCookie, CancellationToken cancellationToken)
         {
+            AssertIsForeground();
+
+            cancellationToken.ThrowIfCancellationRequested();
```

Comment:
Yep, your concern is also my concern. I'd imagine the only times the task gets cancelled are:

1. We remove the document from the background thread while we're looking for it on the UI thread.
2. We had some UI notification (an RDT event, for example) that meant that the work we queued to the UI thread is no longer necessary.

The second doesn't benefit from more cancellations, since both pieces are on the UI thread. And the first seems rare. I wouldn't worry about optimizing it, especially given the risks you called out.